entity-framework/core/what-is-new/ef-core-9.0/whatsnew.md

```diff
@@ -1182,6 +1182,68 @@ protected override void Up(MigrationBuilder migrationBuilder)
 }
 ```
 
+<a name="concurrent-migrations"></a>
+
+### Protection against concurrent migrations
+
+EF9 introduces a locking mechanism when executing migrations to protect against multiple migrations happening simultaneously, which could leave the database in a corrupted state. This should be a rare occurrence when migrations are deployed to the production environment using [recommended methods](/ef/core/managing-schemas/migrations/applying#sql-scripts), however it could happen more often if migrations are applied at runtime using the `IMigrator.Migrate()` method.
+
+> [!NOTE]
+> We recommend not to apply migrations at runtime outside of the development cycle, but in some scenarios it is unavoidable, e.g. [working with using .NET Aspire projects](/dotnet/aspire/database/ef-core-migrations).
+
+<a name="warn-when-no-transaction"></a>
+
+### Warn when multiple migration operations can't be run inside a transaction
+
+Majority of operations performed during migrations are protected by a transaction. This ensures that if for some reason migration fails, the database does not end up in a corrupted state. However, some operations are not wrapped in a transaction (e.g. [operations on memory-optimized tables](/sql/relational-databases/in-memory-oltp/unsupported-sql-server-features-for-in-memory-oltp#scenarios-not-supported), or database altering operations like modifying the database collation). To avoid corrupting the database in case of migration failure, it is recommended that these operations are performed in isolation using a separate migration. EF9 now detects a scenario when a migration contains multiple operations, one of which can't be wrapped in a transaction, and issues a warning.
```

Comment:
```suggestion
The majority of operations performed during migrations are protected by a transaction. This ensures that if for some reason migration fails, the database does not end up in a corrupted state. However, some operations are not wrapped in a transaction (e.g. [operations on memory-optimized tables](/sql/relational-databases/in-memory-oltp/unsupported-sql-server-features-for-in-memory-oltp#scenarios-not-supported), or database-altering operations like modifying the database collation). To avoid corrupting the database in case of a migration failure, it is recommended that these operations are performed in isolation using a separate migration. EF9 now detects a scenario when a migration contains multiple operations, one of which can't be wrapped in a transaction, and issues a warning.
```